kazuki-core-test/src/test/kotlin/com/anaplan/engineering/kazuki/core/test/TestSet.kt

```diff
@@ -1,87 +1,135 @@
-        assertFailsWith<PreconditionFailure> { mk_Set<Int>().arbitrary() }
+        if (allowsEmpty) {
+            causesPreconditionFailure { empty().arbitrary() }
+        }
+        assertEquals(true, create(1, 2, 3).arbitrary() in create(1, 2, 3))
+        assertEquals(false, create(1, 2, 3).arbitrary() in create(4, 5, 6))
     }
 
     @Test
     fun card() {
-        assertEquals(3, mk_Set(1, 2, 3).card)
-        assertEquals(3, mk_Set(mk_Set(1, 2), 2, 3).card)
-        assertEquals(0, mk_Set<Int>().card)
+        if (allowsEmpty) {
+            assertEquals(0, empty().card)
+        }
+        assertEquals(3, create(1, 2, 3).card)
+        assertEquals(3, create(create(1, 2), 2, 3).card)
     }
 
     @Test
     fun dunion() {
-        assertEquals(mk_Set(), dunion(mk_Set<Int>()))
-        assertEquals(mk_Set(1, 2, 3), dunion(mk_Set(mk_Set(1, 2), mk_Set(3))))
-        assertEquals(mk_Set(1, 2, 3), dunion(mk_Set(1, 2, 3)))
-        assertEquals(mk_Set(1), dunion(mk_Set(), mk_Set(1)))
-        assertEquals(mk_Set(mk_Set(1), 2, 3), dunion(mk_Set(mk_Set(1), 2), mk_Set(3)))
-        assertEquals(mk_Set(mk_Set(1, 2), 3), dunion(mk_Set(mk_Set(1, 2), 3)))
+        if (allowsEmpty) {
+            assertEquals(empty(), dunion(empty()))
+            assertEquals(create(1), dunion(empty(), create(1)))
+        }
+        assertEquals(create(1, 2, 3), dunion(mk_Set1(create(1, 2), create(3))))
+        assertEquals(create(1, 2, 3), dunion(create(1, 2, 3)))
+        assertEquals(create(create(1), 2, 3), dunion(mk_Set1(create(1), 2), create(3)))
+        assertEquals(create(create(1, 2), 3), dunion(create(create(1, 2), 3)))
     }
 
     @Test
     fun plus() {
-        assertEquals(mk_Set(1), mk_Set<Int>().plus(1))
-        assertEquals(mk_Set(1), mk_Set(1).plus(mk_Set()))
-        assertEquals(mk_Set(1, 2), mk_Set(1).plus(mk_Set(1, 2)))
-        assertEquals(mk_Set(1, 2), mk_Set(1).plus(2))
+        if (allowsEmpty) {
+            assertEquals(create(1), empty().plus(1))
+            assertEquals(create(1), create(1).plus(empty()))
+        }
+        assertEquals(create(1, 2), create(1).plus(create(1, 2)))
+        assertEquals(create(1, 2), create(1).plus(2))
     }
 
     @Test
     fun minus() {
-        assertEquals(mk_Set(), mk_Set<Int>().minus(1))
-        assertEquals(mk_Set(), mk_Set(1).minus(1))
-        assertEquals(mk_Set(2), mk_Set(1, 2).minus(mk_Set(1)))
-        assertEquals(mk_Set(1), mk_Set(1).minus(mk_Set()))
-        assertEquals(mk_Set(), mk_Set<Int>().minus(mk_Set()))
+        if (allowsEmpty) {
+            assertEquals(empty(), empty().minus(1))
+            assertEquals(empty(), create(1).minus(1))
+            assertEquals(empty(), empty().minus(empty()))
+        } else {
+            causesPreconditionFailure { create(1).minus(1) }
+        }
+        assertEquals(create(2), create(1, 2).minus(create(1)))
+        assertEquals(create(1), create(1).minus(mk_Set()))
```

Comment:
:no_entry_sign: what should happen if we try to take away more than exists?

:pick:  adding an item and adding all the items of a set are two different functions and we should likely separate out cf https://github.com/anaplan-engineering/kazuki/blob/develop/kazuki-core-test/src/test/kotlin/com/anaplan/engineering/kazuki/core/test/TestSequence.kt#L215 and https://github.com/anaplan-engineering/kazuki/blob/develop/kazuki-core-test/src/test/kotlin/com/anaplan/engineering/kazuki/core/test/TestSequence.kt#L231

Reply:
I'll add a test. My understanding is that `A - B` is equivalent to `A union not(B)`, so it should just remove all elements of B that are in A, rather than causing an error, right?

Reply:
It's just regular [set difference](https://en.wikipedia.org/wiki/Complement_(set_theory)#Relative_complement).

I'm not sure what you mean by `not(B)`, but `A - B` cannot be equivalent to `A union X` for any set `X`, unless `A inter B` and `X` are both empty, since `A - B` is always contained in `A`.